crates/tauri-cli/src/remove.rs

```diff
@@ -0,0 +1,118 @@
+// Copyright 2019-2024 Tauri Programme within The Commons Conservancy
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-License-Identifier: MIT
+
+use clap::Parser;
+use colored::Colorize;
+
+use crate::{
+  acl,
+  helpers::{
+    app_paths::{resolve_app_dir, tauri_dir},
+    cargo,
+    npm::PackageManager,
+  },
+  Result,
+};
+
+use std::process::Command;
+
+#[derive(Debug, Parser)]
+#[clap(about = "Remove a tauri plugin from the project")]
+pub struct Options {
+  /// The plugin to remove.
+  pub plugin: String,
+  /// Don't format code with rustfmt
+  #[clap(long)]
+  pub no_fmt: bool,
+}
+
+pub fn command(options: Options) -> Result<()> {
+  crate::helpers::app_paths::resolve();
+  run(options)
+}
+
+pub fn run(options: Options) -> Result<()> {
+  let plugin = options.plugin;
+
+  let plugin_snake_case = plugin.replace('-', "_");
+  let crate_name = format!("tauri-plugin-{plugin}");
+
+  let mut plugins = crate::helpers::plugins::known_plugins();
+  let metadata = plugins.remove(plugin.as_str()).unwrap_or_default();
+
+  let app_dir = resolve_app_dir();
+  let tauri_dir = tauri_dir();
+
+  let target_str = metadata
+    .desktop_only
+    .then_some(r#"cfg(not(any(target_os = "android", target_os = "ios")))"#)
+    .or_else(|| {
+      metadata
+        .mobile_only
+        .then_some(r#"cfg(any(target_os = "android", target_os = "ios"))"#)
+    });
+
+  cargo::uninstall_one(cargo::CargoUninstallOptions {
+    name: &crate_name,
+    cwd: Some(tauri_dir),
+    target: target_str,
+  })?;
+
+  if !metadata.rust_only {
+    if let Some(manager) = app_dir
+      .map(PackageManager::from_project)
+      .and_then(|managers| managers.into_iter().next())
+    {
+      let npm_name = format!("@tauri-apps/plugin-{plugin}");
+      manager.remove(&[npm_name], tauri_dir)?;
+    }
+
+    let _ = acl::permission::rm::command(acl::permission::rm::Options {
+      identifier: format!("{plugin}:default"),
+    });
+  }
+
+  // remove plugin init code from main.rs or lib.rs
+  let plugin_init_fn = if plugin == "stronghold" {
+    "Builder::new(|pass| todo!()).build()"
+  } else if plugin == "localhost" {
+    "Builder::new(todo!()).build()"
+  } else if metadata.builder {
+    "Builder::new().build()"
+  } else {
+    "init()"
+  };
+  let plugin_init = format!(".plugin(tauri_plugin_{plugin_snake_case}::{plugin_init_fn})");
+
+  for file in [tauri_dir.join("src/main.rs"), tauri_dir.join("src/lib.rs")] {
+    let contents = std::fs::read_to_string(&file)?;
+
+    if contents.contains(&plugin_init) {
```

Comment:
this is a naive way to check if the plugin code exists or not, and it doesn't take into account that the user may have formatted their code or not.

Reply:
the naive way is fine for `tauri add` because it is an additive operation not destructive, unlike this `rm` command so we need to be more careful, either by enhancing the detection by 100% or take user input on if the detected line should be removed or not.

Reply:
The only robust way I can think of is to use `syn` to turn the code into an AST that can have the node(s) removed. Would adding that be acceptable?

Reply:
yeah sure, `syn` is fine as well. We use a similar approach to modify js/ts files in `tauri migrate`.

Reply:
it's pretty much impossible to remove all plugin references from the code - I'd just leave that to the user
because there's trait imports, api usage etc

or are we just talking about removing the plugin init() call?

Reply:
@lucasfernog I suppose that is a good point. If we are unlikely to be able to remove everything, should we automagically remove anything?